packages/plugin-workspace-tools/sources/commands/foreach.ts

```diff
@@ -126,14 +134,30 @@ export default class WorkspacesForeachCommand extends BaseCommand {
       ? project.topLevelWorkspace
       : cwdWorkspace!;
 
+    if (configuration.projectCwd === null)
+      throw new UsageError(`This command can only be run from within a Yarn project`);
+
+    const root = this.since ? await gitUtils.fetchRoot(configuration.projectCwd) : null;
+    const base = this.since && root !== null
+      ? await gitUtils.fetchBase(root, {baseRefs: typeof this.since === `string` ? [this.since] : configuration.get(`changesetBaseRefs`)})
+      : null;
+
+    const prospectiveWorkspaces = this.since && root !== null
+      ? Array.from(await gitUtils.fetchChangedWorkspaces(root, {base: base!.hash, project}))
+      : [rootWorkspace, ...(this.from.length > 0 ? rootWorkspace.getRecursiveWorkspaceChildren() : [])];
+
     const fromPredicate = (workspace: Workspace) => micromatch.isMatch(structUtils.stringifyIdent(workspace.locator), this.from);
     const fromCandidates: Array<Workspace> = this.from.length > 0
```

Comment:
Something I found a bit difficult to follow when attempting to implement this behavior myself was juggling the notions of "root" and "base" and "candidate" in order to understand what workspaces are under consideration and why, etc., to which you've added another layer of "prospective". 

I don't have any great suggestions to improve this beyond some tweaking of the names -- factoring each step out to a dedicated function, for instance, probably introduces more confusion than it resolves -- but I wanted to bring it your attention in case you can come up with something. I also think this is a reflection of the somewhat weird interactions between the flags that I talk about in my comment a few lines below.

Reply:
Yeah its definitely feeling like there are some distinct paths here based on what options are being set. But I also feel like it would be harder to keep code in sync and avoid duplication if the logic was forked for each of these paths.

There may be some parts of the code I can pull out into functions or at least include comments to better explain what's going on. I'll have a bit of a think about it for my next commit, but until then any suggestions are welcome.

Reply:
Note that if the logic grows too complex, it's perhaps the sign that it would make more sense to keep it as an external command (and possibly a contrib plugin).

Reply:
I think much of the complexity is because we're grafting on a behavior that isn't _quite_ aligned with the direction the command was going when it was only `--all` or `--from`. I also think there is a future version with slightly different flags that is easier to follow and just as featureful, but would probably require (slight) breaking changes to get there. Banning combining `--since` with `--from`/`--all` should get rid of most of the user-facing weirdness, though maybe not from the implementation.